sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -904,3 +907,49 @@ object RemoveLiteralFromGroupExpressions extends Rule[LogicalPlan] {
       a.copy(groupingExpressions = newGrouping)
   }
 }
+
+/**
+ * For an inner join - remove rows with null keys on both sides
+ */
+object JoinSkewOptimizer extends Rule[LogicalPlan] with PredicateHelper {
+  /**
+   * Adds a null filter on given columns, if any
+   */
+  def addNullFilter(columns: AttributeSet, expr: LogicalPlan): LogicalPlan = {
+    columns.map(IsNotNull(_))
+      .reduceLeftOption(And)
+      .map(Filter(_, expr))
+      .getOrElse(expr)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f@Join(left, right, joinType, joinCondition) =>
+      // get "real" join conditions, which refer both left and right
+      val joinConditionsOnBothRelations = joinCondition
+        .map(splitConjunctivePredicates).getOrElse(Nil)
+        .filter(_.isInstanceOf[EqualTo])
+        .filter(cond => !canEvaluate(cond, left) && !canEvaluate(cond, right))
+
+      def nullableJoinKeys(leftOrRight: LogicalPlan) = {
```

Comment:
so it seems [ExtractEquiJoinKeys](https://github.com/apache/spark/blob/67d468f8d9172569ec9846edc6432240547696dd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala#L167) is the right way to go.